### PR TITLE
Add attachments

### PR DESCRIPTION
**Pull request by nicksellen**
Branch deployment URL: https://add-attachments.dev.karrot.world
Backend PR: https://github.com/karrot-dev/karrot-backend/pull/1256

# Attachments!

- can add them to any conversation message (wall, chat, thread, etc.)
- can add up to 6 attachments, each of up to 10mb
- always uploaded exactly as you give it, no cropping, no resizing
- any file type is supported
- image types will have previews and thumbnails created (1600px and 200px size respectively)
- full screen gallery for any viewable images
- can download, retaining original filename
- upload progress dialog
- this doesn't apply to non-message images (offer images, profile photo, ... those are still as before)

Known bugs:
- ~~gallery view in sidebar chat isn't fullscreen~~ fixed!

TODO:
- [ ] cannot click "download" button on attachments in list, as it's a hover-only thing
- [ ] add a way to use camera on mobile (if it doesn't already work)
- [x] MAKE IT WORK IN PRIVATE CHATS AND APPLICATION CHATS, seems a permission issue to view them...?
- [x] add ellipsis or scrolling when filename is too long
- [ ] ~~BACKEND: use aiofiles.open for sending file, as described here --> https://docs.djangoproject.com/en/4.2/ref/request-response/#fileresponse-objects~~ NO - full request async in django is tricky as middlewares are not all async... used nginx X-Accel-Redirect in the end for faster sending
- [x] BACKEND: maybe add sha256 etag

Probably out of scope next steps:
- allow s3-compatible storage backend
- [x] add headers for caching (and maybe include a content-hash in the path if possible it changes)
- attachment captions
- preview audio/video in browser
- image editing (crop, rotate, ...)
- include attachments (as links?) in email notifications?

attachments

attachments2

gallery

### CODE REVIEW

File: src/messages/components/AttachmentGallery.vue [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+<script setup>
+import {
+  QCarousel,
+  QCarouselSlide,
+  QCarouselControl,
+  QBtn,
+  Platform,
+} from 'quasar'
+import { computed, ref } from 'vue'
+import { useEvent } from 'vue-composable'
```
Comment (nicksellen) [inline src/messages/components/AttachmentGallery.vue:63]: Could use `import { useEventListener } from '@vueuse/core'` instead, as we already have vueuse.

File: src/base/components/UploadProgressDialog.vue [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+      style="width: 300px; margin-bottom: 0;"
+      square
+    >
+      <QLinearProgress
+        v-if="progress"
+        :value="progress.progress"
+        color="pink"
+      />
+      <QCardSection v-if="progress">
+        Uploading {{ humanStorageSize(progress.loaded) }} / {{ humanStorageSize(progress.total) }}
```
Comment (tiltec) [inline src/base/components/UploadProgressDialog.vue:20]: Should be translatable

File: src/messages/components/Attachments.vue [hunk 3]
```diff
@@ -0,0 +1,465 @@
...
+      }
+    case 'text/plain':
+      return {
+        name: 'Text',
+        icon: 'fas fa-file-alt',
+      }
+  }
+  if (contentType.startsWith('image/')) {
+    return {
+      name: 'Image',
```
Comment (tiltec) [inline src/messages/components/Attachments.vue:402]: could be translatable, also "Text" above and "Audio" below

File: src/messages/components/ConversationCompose.vue [hunk 4]
```diff
@@ -83,12 +89,15 @@
           :prevent-white-space="false"
           small
         />
+        <pre>images: {{ message.images }}</pre>
```
Comment (tiltec) [inline src/messages/components/ConversationCompose.vue:92]: debug statement could be removed

File: src/messages/components/Attachments.vue [hunk 5]
```diff
@@ -0,0 +1,465 @@
...
+
+import { showToast } from '@/utils/toasts'
+import { isViewableImageContentType } from '@/utils/utils'
+
+import AttachmentGallery from '@/messages/components/AttachmentGallery.vue'
+
+const { humanStorageSize } = format
+
+const MAX_ATTACHMENT_COUNT = 6
+const MAX_ATTACHMENT_SIZE = 10 * 1024 * 1024
```
Comment (tiltec) [inline src/messages/components/Attachments.vue:123]: out of sync with backend, maybe could be part of bootstrap/config endpoint?

File: src/messages/components/AttachmentGallery.vue (hunk 1 above)
Comment (tiltec) [reply src/messages/components/AttachmentGallery.vue:63]: Seems good to avoid another dependency.

**Review by tiltec — APPROVED** [review]
Great! I'm sure the groups will find it useful!

File: src/messages/components/Attachments.vue (hunk 5 above)
Comment (nicksellen) [reply src/messages/components/Attachments.vue:123]: done, added to backend :)

**Pull request merged**